<h2>FIN5622 Project #2
    <p class="text-primary">Loan Default Prediction with Logistic Regression</p>
</h2>

<div class = "panel panel-success">
    <div class = "panel-heading"> 
    <span style="color:black"><b>Project Introduction</b>
    </div>
    <div class = "panel-body"> 
    In this project we will build a logistic regression model to predict borrower defaults on a set of historical loan performance dataset.
    </div>
</div>

<div class = "panel panel-info">
    <div class = "panel-heading"> 
    <span style="color:black"><b>Instructions</b>
    </div>
    <div class = "panel-body"> 
    Please code you solutions directly in this Notebook under each task in the cells provided below. Tasks are intended to be done in order with results from previous tasks informing work in later tasks. Include all code used to produce your results, and support your answers with comments if neccessary. <b>Any written text must be typed in a markdown cell to receive credit!</b> You may feel free to add additional Code and/or Markdown cells as needed. Code must be written using <b>Python 3</b>.
    <br><br>Please order the cells in your Notebook in the order in which they are intended to run! It is expected that your Notebook is able to run in sequence from start to finish without any errors (before submitting check by: <code>Kernel > Restart & Run All</code>).
    <br><br>Once ready, please submit your assignment as an <b>.ipynb file</b> (<code>File > Download as > Notebook</code>). You must name your files in the following format: <b><code>FIN5622_Project_2_Richard_Lee_tuc12345.ipynb</code></b>.
    </div>
</div>

---
<div class = "alert alert-block alert-warning">
    <b>Note</b>: Import all libraries that you plan on using in the code block below.
</div>

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix
from sklearn.metrics import auc, roc_curve, confusion_matrix

<div class = "panel panel-primary">
    <div class = "panel-heading"> 
    <span style="color:white"><b>Task 1</b>
    </div>
    <div class = "panel-body"> 
    Read in the dataset "loan_performance_data_2016.csv" into a dataframe called <code>loans</code> using the pandas <code>pd.read_csv()</code> function. Check that there are 405,180 loans present in the dataframe.
    <br><br>
    Note that there are 16 feature columns and 1 target column (i.e. "loan_default") in this dataset. Please review the file "Loan_Data_Dictionary.xlsx" for definitions on each column. Pick a variable that you believe will be an important explanatory variable for loan defaults and <b>explain why</b>.
    <p style="text-align:right;">[5pts]</p>
    </div>
</div>

In [2]:
loans = pd.read_csv('loan_performance_data_2016.csv')

In [3]:
loans

loan_amnt  loan_term  int_rate  fico    dti  annual_income  \
0           7.200        3.0      5.32   750  19.05           49.0   
1          12.000        5.0     11.99   670   6.59           49.0   
2          11.425        3.0     19.53   730  35.56           26.0   
3          17.500        3.0     13.67   665  38.29           46.0   
4          22.875        5.0     17.27   670  11.75           65.0   
...           ...        ...       ...   ...    ...            ...   
405175     16.000        5.0     12.79   665  12.25          150.0   
405176     24.000        5.0     12.79   665  19.61           95.0   
405177     40.000        5.0     10.49   705  12.75          227.0   
405178     24.000        5.0     14.49   660  18.30          110.0   
405179     14.000        5.0     14.49   660  23.36           95.0   

        income_verification  home_mortgage  home_own  home_rent  \
0                         0              1         0          0   
1                         0              1         0          0   
2                         0              0         0          1   
3                         0              0         0          1   
4                         0              0         0          1   
...                     ...            ...       ...        ...   
405175                    0              0         0          1   
405176                    0              1         0          0   
405177                    1              1         0          0   
405178                    0              0         0          1   
405179                    1              1         0          0   

        debt_consolidation  coborrower  num_accounts  revol_bal  revol_util  \
0                        1           0            11      9.309        18.4   
1                        0           0            10     12.152        50.6   
2                        0           0            11      2.096         7.2   
3                        1           0            33     18.847        62.6   
4                        1           0             6     25.200        90.0   
...                    ...         ...           ...        ...         ...   
405175                   0           0            12      7.700        55.0   
405176                   0           0             5     49.431        84.4   
405177                   0           0             5      8.633        64.9   
405178                   1           0            10     17.641        68.1   
405179                   0           0             8      7.662        54.0   

        derogatory_records  loan_default  
0                        0             0  
1                        0             1  
2                        0             0  
3                        0             0  
4                        0             0  
...                    ...           ...  
405175                   1             0  
405176                   0             0  
405177                   0             0  
405178                   1             1  
405179                   0             0  

[405180 rows x 17 columns]

A variable that I think would be an important for loan defaults would be annual income.  This is very important as annual income can be used to relatively establish how much money the home buyer would be able to set aside to pay the loan (ie someone with an annual income of 400,000 dollars would be able to set aside more money than someone who makes 100,000 dollars annually for a house).  This also makes it so that the higher the income the more likely it is that the buyer would not default on the loan as they would have less of a monetary reason to default on a loan. 

<div class = "panel panel-primary">
    <div class = "panel-heading"> 
    <span style="color:white"><b>Task 2</b>
    </div>
    <div class = "panel-body"> 
    We will be creating a logistic regression model to predict the "loan_default" column, with the rest of the columns as feature variables. 
    <br><br>
    Create a new variable called <code>loan_features</code> to store just the columns that are input variables, and another variable called <code>loan_target</code> to store the prediction target. Each array shoud have the following shapes: 
    <ul>
        <li><code>loan_features.shape = (405180, 16)</code></li>
        <li><code>loan_target.shape = (405180,)</code></li>
    </ul>
    <p style="text-align:right;">[10pts]</p>
    </div>
</div>

In [4]:
loans.columns

Index(['loan_amnt', 'loan_term', 'int_rate', 'fico', 'dti', 'annual_income',
       'income_verification', 'home_mortgage', 'home_own', 'home_rent',
       'debt_consolidation', 'coborrower', 'num_accounts', 'revol_bal',
       'revol_util', 'derogatory_records', 'loan_default'],
      dtype='object')

In [5]:
loans
loan_target = ['loan_default']
loan_features = loans.drop('loan_default', axis =1).columns.tolist()
loan_features, loan_target


(['loan_amnt',
  'loan_term',
  'int_rate',
  'fico',
  'dti',
  'annual_income',
  'income_verification',
  'home_mortgage',
  'home_own',
  'home_rent',
  'debt_consolidation',
  'coborrower',
  'num_accounts',
  'revol_bal',
  'revol_util',
  'derogatory_records'],
 ['loan_default'])

In [6]:
X_train, X_test, y_train, y_test = train_test_split(
    loans[loan_features], loans[loan_target], test_size = 0.2, random_state = 5622)

In [7]:
loan_features

['loan_amnt',
 'loan_term',
 'int_rate',
 'fico',
 'dti',
 'annual_income',
 'income_verification',
 'home_mortgage',
 'home_own',
 'home_rent',
 'debt_consolidation',
 'coborrower',
 'num_accounts',
 'revol_bal',
 'revol_util',
 'derogatory_records']

In [8]:
loans[loan_features].shape

(405180, 16)

In [9]:
loans[loan_target].shape

(405180, 1)

<div class = "panel panel-primary">
    <div class = "panel-heading"> 
    <span style="color:white"><b>Task 3</b>
    </div>
    <div class = "panel-body"> 
    Fit a logistic regression model using the data stored in <code>loan_features</code> and <code>loan_target</code>. 
    <br><br>
    Recall that this is done by calling <code>LogisticRegression()</code> from the <code>sklearn.linear_model</code> module, and using the <code>.fit()</code> method. Store the fitted model in a variable called <code>logistic_model</code>.
    <p style="text-align:right;">[15pts]</p>
    </div>
</div>

In [10]:
logistic_model = LogisticRegression(solver = 'liblinear').fit(X = loans[loan_features], y = loans[loan_target])

C:\Users\15134\anaconda3\lib\site-packages\sklearn\utils\validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)


In [11]:
logistic_model.intercept_, logistic_model.coef_

(array([0.0061452]),
 array([[ 0.01049985,  0.31148077,  0.09270302, -0.00590638,  0.00987407,
         -0.00118002,  0.06060728, -0.16932907,  0.00500468,  0.17051501,
         -0.04210274, -0.02461696,  0.0128059 , -0.00430041, -0.0007566 ,
          0.06090127]]))

<div class = "panel panel-primary">
    <div class = "panel-heading"> 
    <span style="color:white"><b>Task 4</b>
    </div>
    <div class = "panel-body"> 
    Examine the model's fitted coefficients (stored in the <code>coef_</code> attribute), and answer the following questions:
    <ul>
        <li>How should we interpret the output of the logistic regression model?</li>
        <li>How should we interpret a positive coefficient in the logistic regression model? How about a negative coefficient?</li>
        <li>Are all the signs of the fitted coeffiecients in your logistic regression model for each variable as expected? If not, list the variables with unexpected signs and explain what the expected sign on the coefficient should be and why.</li>
    </ul>
    <p style="text-align:right;">[20pts]</p>
    </div>
</div>

In [12]:
logistic_model.coef_

array([[ 0.01049985,  0.31148077,  0.09270302, -0.00590638,  0.00987407,
        -0.00118002,  0.06060728, -0.16932907,  0.00500468,  0.17051501,
        -0.04210274, -0.02461696,  0.0128059 , -0.00430041, -0.0007566 ,
         0.06090127]])

In [13]:
logistic_model.intercept_

array([0.0061452])

In [14]:
pd.Series(logistic_model.coef_.squeeze(), index = loan_features)

loan_amnt              0.010500
loan_term              0.311481
int_rate               0.092703
fico                  -0.005906
dti                    0.009874
annual_income         -0.001180
income_verification    0.060607
home_mortgage         -0.169329
home_own               0.005005
home_rent              0.170515
debt_consolidation    -0.042103
coborrower            -0.024617
num_accounts           0.012806
revol_bal             -0.004300
revol_util            -0.000757
derogatory_records     0.060901
dtype: float64

The outputs of the logistic regression model are the coefficeints of each of the columns in the dataset that was brought in with the readcsv function

a positive coefficent indicates a positive relationship between the explanatory variable and the probability of belonging to the target class, while a negative coefficient indicates a negative relationship between the explanatory variable and the probability of it belonging to the target class. 

The signs of the fitted coefficients in the logistic regression model are as expected. 

<div class = "panel panel-primary">
    <div class = "panel-heading"> 
    <span style="color:white"><b>Task 5</b>
    </div>
    <div class = "panel-body"> 
    Calculate the logistic regression model's classification error on our dataset <code>loan_features</code> and <code>loan_target</code>. Recall that this is done by calling the <code>.score()</code> method from <code>logistic_model</code>. The AUC should be approximately 80%.
    <p style="text-align:right;">[10pts]</p>
    </div>
</div>

In [15]:
logistic_model.score(X = loans[loan_features], y = loans[loan_target])

0.8062737548743768

<div class = "panel panel-primary">
    <div class = "panel-heading"> 
    <span style="color:white"><b>Task 6</b>
    </div>
    <div class = "panel-body"> 
    Use the fitted logistic regression model saved in <code>logistic_model</code> to predict the loan default probabilities for the loans in <code>loan_features</code>. Recall that this is done by calling the <code>.predict_proba()</code> method from logistic_model. 
    <br><br>
    Note that while the output of <code>.predict_proba()</code> is a 2-d array, we only need the 2nd column, which is the probability of loan defaults, and should be a vector of numbers between 0 and 1. Index the 2nd columns and store the probabilities in a variable called <code>model_prob</code>.
    <br><br>
    Calculate the average predicted probability of default for the loans in <code>loan_features</code>, along with the average actual default probability observed in <code>loan_target</code>. Both should be approximately equalt to 19.6%.
    <p style="text-align:right;">[15pts]</p>
    </div>
</div>

In [16]:
logistic_model.predict_proba(loans[loan_features])

array([[0.94856619, 0.05143381],
       [0.78258224, 0.21741776],
       [0.694927  , 0.305073  ],
       ...,
       [0.81470655, 0.18529345],
       [0.62211083, 0.37788917],
       [0.69123899, 0.30876101]])

In [17]:
model_prob = logistic_model.predict_proba(loans[loan_features])[:, 1]
model_prob

array([0.05143381, 0.21741776, 0.305073  , ..., 0.18529345, 0.37788917,
       0.30876101])

In [18]:
avg_prob_features = np.average(logistic_model.predict_proba(loans[loan_features])[:, 1])
avg_prob_features

0.19596379178403794

In [19]:
avg_prob_target = np.average(loans['loan_default'])
avg_prob_target

0.19609803050496075

<div class = "panel panel-primary">
    <div class = "panel-heading"> 
    <span style="color:white"><b>Task 7</b>
    </div>
    <div class = "panel-body"> 
    Calculate the loan default predictions from the logistic regression model on <code>loan_features</code> by using the <code>.predict()</code> method. Note that the results should be a vector of 1's and 0's. Store the predictions in a variable called <code>model_pred</code>. 
    <br><br>
    Explain the difference between the functions <code>.predict_proba()</code> and <code>.predict()</code>.
    <p style="text-align:right;">[10pts]</p>
    </div>
</div>

In [20]:
model_pred = logistic_model.predict(loans[loan_features])
model_pred.mean()

0.03774865491880152

In [21]:
model_pred

array([0, 0, 0, ..., 0, 0, 0], dtype=int64)

In [22]:
model_prob

array([0.05143381, 0.21741776, 0.305073  , ..., 0.18529345, 0.37788917,
       0.30876101])

.predict() is used to give values such as 0 and 1, and .predict_proba() is used to give values in probabilities. 

<div class = "panel panel-primary">
    <div class = "panel-heading"> 
    <span style="color:white"><b>Task 8</b>
    </div>
    <div class = "panel-body"> 
    Create the confusion matrix for the predictions in <code>model_pred</code> compared to the actual outcomes in <code>loan_target</code>. Recall that this can be done using the function <code>confusion_matrix()</code> from the <code>sklearn.metrics</code> module.
    <br><br>
    Interpret and explain the results of each of the 4 cells of the confusion matrix in the context of this project. 
    <p style="text-align:right;">[15pts]</p>
    </div>
</div>

In [23]:
confusion_matrix(loans[loan_target], model_pred) / loans[loan_target].shape[0] 

array([[0.78621353, 0.01768843],
       [0.17603781, 0.02006022]])

In the context of the project, the left top cell (0.78621353) indicates the true positive of this dataset.  0.1768843 in the top left indicates the false positive.  0.17603781 in the bottom left indicates the false negative, while 0.02006022 in the bottom right indicates the true negative. 